src/main/scala/com/wavesplatform/matcher/MatcherSnapshotStore.scala

```diff
@@ -0,0 +1,204 @@
+package com.wavesplatform.matcher
+
+import java.io._
+import java.net.{URLDecoder, URLEncoder}
+import java.nio.ByteBuffer
+
+import akka.persistence._
+import akka.persistence.serialization._
+import akka.persistence.snapshot.SnapshotStore
+import akka.serialization.SerializationExtension
+import com.google.common.base.Charsets.UTF_8
+import com.google.common.primitives.{Bytes, Ints}
+import com.typesafe.config.Config
+import com.wavesplatform.database._
+import com.wavesplatform.db.openDB
+import org.iq80.leveldb.ReadOptions
+import scorex.utils.ScorexLogging
+
+import scala.concurrent.Future
+import scala.util._
+import scala.util.control.NonFatal
+
+class MatcherSnapshotStore(config: Config) extends SnapshotStore {
+  import MatcherSnapshotStore._
+
+  private val streamDispatcher = context.system.dispatchers.lookup(config.getString("stream-dispatcher"))
+
+  private val serializationExtension = SerializationExtension(context.system)
+
+  private val writableDB = openDB(config.getString("leveldb-dir"), 100)
+
+  private def readOnly[A](f: ReadOnlyDB => A): A = {
+    val s = writableDB.getSnapshot
+    try f(new ReadOnlyDB(writableDB, new ReadOptions().snapshot(s)))
+    finally s.close()
+  }
+
+  private def readWrite[A](f: RW => A): A = {
+    val rw = new RW(writableDB)
+    try f(rw)
+    finally rw.close()
+  }
+
+  private var pendingActions = Map.empty[String, Future[_]]
+
+  private def nextAction[A](persistenceId: String, f: => A): Future[A] = {
+    val newAction = pendingActions
+      .getOrElse(persistenceId, Future(())(streamDispatcher))
```

Comment:
Small tip: `Future.successful` is better in such situations, because it doesn't run any computation. Also it doesn't require a scheduler